In [2]:
import pandas as pd
import matplotlib as plt

In [25]:
#carga del arhivo movies.

df = pd.read_csv("movies_dataset.csv")
df.head()

C:\Users\fmgc1\AppData\Local\Temp\ipykernel_13924\215091187.py:3: DtypeWarning: Columns (10) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv("movies_dataset.csv")


adult                              belongs_to_collection    budget  \
0  False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1  False                                                NaN  65000000   
2  False  {'id': 119050, 'name': 'Grumpy Old Men Collect...         0   
3  False                                                NaN  16000000   
4  False  {'id': 96871, 'name': 'Father of the Bride Col...         0   

                                              genres  \
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   
2  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...   
3  [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...   
4                     [{'id': 35, 'name': 'Comedy'}]   

                               homepage     id    imdb_id original_language  \
0  http://toystory.disney.com/toy-story    862  tt0114709                en   
1                                   NaN   8844  tt0113497                en   
2                                   NaN  15602  tt0113228                en   
3                                   NaN  31357  tt0114885                en   
4                                   NaN  11862  tt0113041                en   

                original_title  \
0                    Toy Story   
1                      Jumanji   
2             Grumpier Old Men   
3            Waiting to Exhale   
4  Father of the Bride Part II   

                                            overview  ... release_date  \
0  Led by Woody, Andy's toys live happily in his ...  ...   1995-10-30   
1  When siblings Judy and Peter discover an encha...  ...   1995-12-15   
2  A family wedding reignites the ancient feud be...  ...   1995-12-22   
3  Cheated on, mistreated and stepped on, the wom...  ...   1995-12-22   
4  Just when George Banks has recovered from his ...  ...   1995-02-10   

       revenue runtime                                   spoken_languages  \
0  373554033.0    81.0           [{'iso_639_1': 'en', 'name': 'English'}]   
1  262797249.0   104.0  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...   
2          0.0   101.0           [{'iso_639_1': 'en', 'name': 'English'}]   
3   81452156.0   127.0           [{'iso_639_1': 'en', 'name': 'English'}]   
4   76578911.0   106.0           [{'iso_639_1': 'en', 'name': 'English'}]   

     status                                            tagline  \
0  Released                                                NaN   
1  Released          Roll the dice and unleash the excitement!   
2  Released  Still Yelling. Still Fighting. Still Ready for...   
3  Released  Friends are the people who let you be yourself...   
4  Released  Just When His World Is Back To Normal... He's ...   

                         title  video vote_average vote_count  
0                    Toy Story  False          7.7     5415.0  
1                      Jumanji  False          6.9     2413.0  
2             Grumpier Old Men  False          6.5       92.0  
3            Waiting to Exhale  False          6.1       34.0  
4  Father of the Bride Part II  False          5.7      173.0  

[5 rows x 24 columns]

De un primer analisis rapido y consultando el diccionario de datos pudimos identificar algunas columnas que consideramos que no aportarian informacion util al algoritmo de recomendacion. Por ejemplo la columna homepage contiene las direcciones web de las paginas de las peliculas, la columna overview contiene una reseña del contenido de la pelicula.

In [26]:
#Eliminamos las columnas que consideramos que no aportarian informacion util
# El resto se hara en el apartado de TRANSFORMACIONES mas abajo.
df = df.drop(columns=['overview', 'status', 'tagline'])
df.head()

adult                              belongs_to_collection    budget  \
0  False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1  False                                                NaN  65000000   
2  False  {'id': 119050, 'name': 'Grumpy Old Men Collect...         0   
3  False                                                NaN  16000000   
4  False  {'id': 96871, 'name': 'Father of the Bride Col...         0   

                                              genres  \
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   
2  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...   
3  [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...   
4                     [{'id': 35, 'name': 'Comedy'}]   

                               homepage     id    imdb_id original_language  \
0  http://toystory.disney.com/toy-story    862  tt0114709                en   
1                                   NaN   8844  tt0113497                en   
2                                   NaN  15602  tt0113228                en   
3                                   NaN  31357  tt0114885                en   
4                                   NaN  11862  tt0113041                en   

                original_title popularity  ...  \
0                    Toy Story  21.946943  ...   
1                      Jumanji  17.015539  ...   
2             Grumpier Old Men    11.7129  ...   
3            Waiting to Exhale   3.859495  ...   
4  Father of the Bride Part II   8.387519  ...   

                                production_companies  \
0     [{'name': 'Pixar Animation Studios', 'id': 3}]   
1  [{'name': 'TriStar Pictures', 'id': 559}, {'na...   
2  [{'name': 'Warner Bros.', 'id': 6194}, {'name'...   
3  [{'name': 'Twentieth Century Fox Film Corporat...   
4  [{'name': 'Sandollar Productions', 'id': 5842}...   

                                production_countries release_date  \
0  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-10-30   
1  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-12-15   
2  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-12-22   
3  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-12-22   
4  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-02-10   

       revenue  runtime                                   spoken_languages  \
0  373554033.0     81.0           [{'iso_639_1': 'en', 'name': 'English'}]   
1  262797249.0    104.0  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...   
2          0.0    101.0           [{'iso_639_1': 'en', 'name': 'English'}]   
3   81452156.0    127.0           [{'iso_639_1': 'en', 'name': 'English'}]   
4   76578911.0    106.0           [{'iso_639_1': 'en', 'name': 'English'}]   

                         title  video vote_average  vote_count  
0                    Toy Story  False          7.7      5415.0  
1                      Jumanji  False          6.9      2413.0  
2             Grumpier Old Men  False          6.5        92.0  
3            Waiting to Exhale  False          6.1        34.0  
4  Father of the Bride Part II  False          5.7       173.0  

[5 rows x 21 columns]

In [51]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45466 entries, 0 to 45465
Data columns (total 21 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   adult                  45466 non-null  object 
 1   belongs_to_collection  4494 non-null   object 
 2   budget                 45466 non-null  object 
 3   genres                 45466 non-null  object 
 4   homepage               7782 non-null   object 
 5   id                     45466 non-null  object 
 6   imdb_id                45449 non-null  object 
 7   original_language      45455 non-null  object 
 8   original_title         45466 non-null  object 
 9   popularity             45461 non-null  object 
 10  poster_path            45080 non-null  object 
 11  production_companies   45463 non-null  object 
 12  production_countries   45463 non-null  object 
 13  release_date           45379 non-null  object 
 14  revenue                45460 non-null  float64
 15  ru

Notamos que de de las 45466 filas totales la columna title contiene 6 valores nulos, entonces los analizamos

In [52]:
df[df['title'].isna()]

adult  \
19729                                              False   
19730                                 - Written by Ørnås   
29502                                              False   
29503   Rune Balot goes to a casino connected to the ...   
35586                                              False   
35587   Avalanche Sharks tells the story of a bikini ...   

                                   belongs_to_collection  \
19729                                                NaN   
19730                                           0.065736   
29502  {'id': 122661, 'name': 'Mardock Scramble Colle...   
29503                                           1.931659   
35586                                                NaN   
35587                                           2.185485   

                                 budget  \
19729                                 0   
19730  /ff9qCepilowshEtG2GYWwzt2bs4.jpg   
29502                                 0   
29503  /zV8bHuSL6WXoD6FWogP9j4x80bL.jpg   
35586                                 0   
35587  /zaSf5OG7V8X8gqFvly88zDdRm46.jpg   

                                                  genres  \
19729  [{'id': 28, 'name': 'Action'}, {'id': 53, 'nam...   
19730  [{'name': 'Carousel Productions', 'id': 11176}...   
29502  [{'id': 16, 'name': 'Animation'}, {'id': 878, ...   
29503  [{'name': 'Aniplex', 'id': 2883}, {'name': 'Go...   
35586  [{'id': 10770, 'name': 'TV Movie'}, {'id': 28,...   
35587  [{'name': 'Odyssey Media', 'id': 17161}, {'nam...   

                                                homepage          id  \
19729                                                NaN       82663   
19730  [{'iso_3166_1': 'CA', 'name': 'Canada'}, {'iso...  1997-08-20   
29502                      http://m-scramble.jp/exhaust/      122662   
29503  [{'iso_3166_1': 'US', 'name': 'United States o...  2012-09-29   
35586                                                NaN      249260   
35587           [{'iso_3166_1': 'CA', 'name': 'Canada'}]  2014-01-01   

         imdb_id original_language                            original_title  \
19729  tt0113002                en                              Midnight Man   
19730          0             104.0  [{'iso_639_1': 'en', 'name': 'English'}]   
29502  tt2423504                ja                          マルドゥック・スクランブル 排気   
29503          0              68.0      [{'iso_639_1': 'ja', 'name': '日本語'}]   
35586  tt2622826                en                          Avalanche Sharks   
35587          0              82.0  [{'iso_639_1': 'en', 'name': 'English'}]   

                  popularity  ... production_companies production_countries  \
19729                    NaN  ...                  NaN                  NaN   
19730                    NaN  ...                False                  6.0   
29502                    NaN  ...                  NaN                  NaN   
29503                    NaN  ...                False                  7.0   
35586                    NaN  ...                  NaN                  NaN   
35587  Beware Of Frost Bites  ...                False                  4.3   

      release_date revenue  runtime  spoken_languages title video  \
19729          NaN     NaN      NaN               NaN   NaN   NaN   
19730            1     NaN      NaN               NaN   NaN   NaN   
29502          NaN     NaN      NaN               NaN   NaN   NaN   
29503           12     NaN      NaN               NaN   NaN   NaN   
35586          NaN     NaN      NaN               NaN   NaN   NaN   
35587           22     NaN      NaN               NaN   NaN   NaN   

      vote_average  vote_count  
19729          NaN         NaN  
19730          NaN         NaN  
29502          NaN         NaN  
29503          NaN         NaN  
35586          NaN         NaN  
35587          NaN         NaN  

[6 rows x 21 columns]

Notamos que podemos eliminar del dataset estas 6 filas

In [27]:
# elimino las filas cuyo valores en la columna title es NAN
df = df.dropna(subset=['title'])

Analizamos ahora la columna adult

In [6]:
df['adult'].value_counts()

adult
False    45451
True         9
Name: count, dtype: int64

Notamos 9 filas con valor TRUE las analizamos en detalle

In [55]:
df[df['adult'] == "True"]

adult belongs_to_collection  budget  \
19489  True                   NaN       0   
28701  True                   NaN       0   
31934  True                   NaN       0   
32113  True                   NaN       0   
39901  True                   NaN       0   
39902  True                   NaN       0   
40574  True                   NaN       0   
41009  True                   NaN  750000   
43090  True                   NaN       0   

                                                  genres  \
19489                     [{'id': 27, 'name': 'Horror'}]   
28701                      [{'id': 80, 'name': 'Crime'}]   
31934                     [{'id': 35, 'name': 'Comedy'}]   
32113  [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...   
39901  [{'id': 80, 'name': 'Crime'}, {'id': 27, 'name...   
39902  [{'id': 35, 'name': 'Comedy'}, {'id': 10402, '...   
40574  [{'id': 80, 'name': 'Crime'}, {'id': 27, 'name...   
41009  [{'id': 53, 'name': 'Thriller'}, {'id': 80, 'n...   
43090  [{'id': 878, 'name': 'Science Fiction'}, {'id'...   

                                   homepage      id    imdb_id  \
19489                                   NaN    5422  tt0079642   
28701  http://www.imdb.com/title/tt0119931/  169158  tt0119931   
31934                                   NaN   44781  tt0322232   
32113             http://www.dietofsex.com/  324230  tt3094816   
39901                                   NaN   35731  tt1161951   
39902                                   NaN   55774  tt1153101   
40574                                   NaN   39922  tt0070696   
41009                                   NaN  376004  tt4044464   
43090                                   NaN  448290  tt2697688   

      original_language                       original_title popularity  ...  \
19489                it  Le notti erotiche dei morti viventi   1.420104  ...   
28701                en                             Standoff   1.349745  ...   
31934                cn                                發電悄嬌娃   0.001803  ...   
32113                es                          Diet of Sex   6.889832  ...   
39901                en           Amateur Porn Star Killer 2   0.587751  ...   
39902                en                             The Band   0.417214  ...   
40574                da                              Dværgen   1.182635  ...   
41009                en                           Adulterers    1.13377  ...   
43090                en                           Half -Life    0.28268  ...   

                                    production_companies  \
19489             [{'name': 'Stefano Film', 'id': 6117}]   
28701         [{'name': 'Trimark Pictures', 'id': 4063}]   
31934                                                 []   
32113              [{'name': 'KeyAtomics', 'id': 46055}]   
39901         [{'name': 'Alter Ego Cinema', 'id': 5639}]   
39902                                                 []   
40574           [{'name': 'Scandica Film', 'id': 48812}]   
41009                                                 []   
43090  [{'name': 'Bad Robot', 'id': 11461}, {'name': ...   

                                    production_countries release_date revenue  \
19489            [{'iso_3166_1': 'IT', 'name': 'Italy'}]   1980-11-18     0.0   
28701  [{'iso_3166_1': 'US', 'name': 'United States o...   1998-01-11     0.0   
31934        [{'iso_3166_1': 'HK', 'name': 'Hong Kong'}]   2001-04-26     0.0   
32113            [{'iso_3166_1': 'ES', 'name': 'Spain'}]   2014-02-14     0.0   
39901  [{'iso_3166_1': 'US', 'name': 'United States o...   2008-05-13     0.0   
39902        [{'iso_3166_1': 'AU', 'name': 'Australia'}]   2009-11-17     0.0   
40574  [{'iso_3166_1': 'DK', 'name': 'Denmark'}, {'is...   1973-12-20     0.0   
41009                                                 []   2016-01-05     0.0   
43090  [{'iso_3166_1': 'US', 'name': 'United States o...          NaN     0.0   

       runtime                            spoken_languages  \
19489    112.0   [

se decide hacer un recomendador "family friendly" por lo que nos quedamos solo con los titulos que no son para adultos

In [31]:
df = df[df['adult'] == "False"]

Una vez echo esto la columna adult queda con todos los valores "False" por lo que se puede quitar esta columna

In [32]:
#comprobamos
df['adult'].value_counts()

adult
False    45451
Name: count, dtype: int64

In [24]:
#Se hara mas adelante en el apartado TRANSFORMACIONES
#eliminamos la columna
#df = df.drop('adult', axis=1)

## Transformaciones

1_ Eliminar las columnas innecesarias para el desarrollo de la API, como "video", "imdb_id", "adult", "original_title", "vote_count", "poster_path" y "homepage".

2_ Rellenar con 0 los registros con valores nulos en las columnas "revenue" y "budget".

3_ Eliminar los registros que contengan valores nulos en la columna "release date".

4_ Asegurarse de que la columna "release_year" esté en formato AAAA-MM-DD y generar una nueva columna que contenga solo el año de estreno.

5_ Crear una nueva columna "return" que extraiga el retorno de inversión de cada registro, dividiendo la columna "revenue" entre la columna "budget".

6_ Desanidar los datos que se encuentran en formato de listas de diccionarios para poder unirlos nuevamente a la base de datos de manera separada.



In [34]:
#1 Eliminamos columnas inecesarias:
df = df.drop(columns=['video', 'imdb_id','adult', 'original_title','vote_count', 'poster_path','homepage'])
df.head()


belongs_to_collection    budget  \
0  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1                                                NaN  65000000   
2  {'id': 119050, 'name': 'Grumpy Old Men Collect...         0   
3                                                NaN  16000000   
4  {'id': 96871, 'name': 'Father of the Bride Col...         0   

                                              genres     id original_language  \
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...    862                en   
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   8844                en   
2  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...  15602                en   
3  [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...  31357                en   
4                     [{'id': 35, 'name': 'Comedy'}]  11862                en   

  popularity                               production_companies  \
0  21.946943     [{'name': 'Pixar Animation Studios', 'id': 3}]   
1  17.015539  [{'name': 'TriStar Pictures', 'id': 559}, {'na...   
2    11.7129  [{'name': 'Warner Bros.', 'id': 6194}, {'name'...   
3   3.859495  [{'name': 'Twentieth Century Fox Film Corporat...   
4   8.387519  [{'name': 'Sandollar Productions', 'id': 5842}...   

                                production_countries release_date  \
0  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-10-30   
1  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-12-15   
2  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-12-22   
3  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-12-22   
4  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-02-10   

       revenue  runtime                                   spoken_languages  \
0  373554033.0     81.0           [{'iso_639_1': 'en', 'name': 'English'}]   
1  262797249.0    104.0  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...   
2          0.0    101.0           [{'iso_639_1': 'en', 'name': 'English'}]   
3   81452156.0    127.0           [{'iso_639_1': 'en', 'name': 'English'}]   
4   76578911.0    106.0           [{'iso_639_1': 'en', 'name': 'English'}]   

                         title  vote_average  
0                    Toy Story           7.7  
1                      Jumanji           6.9  
2             Grumpier Old Men           6.5  
3            Waiting to Exhale           6.1  
4  Father of the Bride Part II           5.7

In [43]:
#2_ Rellenar con 0 los registros con valores nulos en las columnas "revenue" y "budget".

df["budget"] = df["budget"].fillna(0)
df["revenue"] = df["revenue"].fillna(0)

In [49]:
df2 = df.copy()

In [50]:
#3_ Eliminar los registros que contengan valores nulos en la columna "release date".
df.dropna(subset=['release_date'], inplace=True)

In [56]:
#4_ Asegurarse de que la columna "release_year" esté en formato AAAA-MM-DD y generar una nueva columna que contenga solo el año de estreno.
df['release_date'] = pd.to_datetime(df['release_date'], errors='coerce')
df['release_year'] = df['release_date'].dt.year

In [59]:
#comprobamos la nueva columna con solo el año de estreno
df['release_year']

0        1995
1        1995
2        1995
3        1995
4        1995
         ... 
45460    1991
45462    2011
45463    2003
45464    1917
45465    2017
Name: release_year, Length: 45368, dtype: int32

In [67]:
# 5_ Crear una nueva columna "return" que extraiga el retorno de inversión de cada registro, 
# dividiendo la columna "revenue" entre la columna "budget".

#primero pasamos a numeric la columna Budget que figura como String
df['budget'] = pd.to_numeric(df['budget'], errors='coerce')

#creamos la nueva columna 
df['return'] = df['revenue'] / df['budget']
df['return']


0        12.451801
1         4.043035
2              NaN
3         5.090760
4              inf
           ...    
45460          NaN
45462          NaN
45463          NaN
45464          NaN
45465          NaN
Name: return, Length: 45368, dtype: float64

In [73]:
df2 = df.copy()

In [74]:
# 6_ Desanidar los datos que se encuentran en formato de listas de diccionarios 
# para poder unirlos nuevamente a la base de datos de manera separada.

#creamos la funcion de convertir a diccionario el texto
def str_to_dict(text):
  try:
    return eval(text)
  except:
    return {}

#aplicamos la funcion a la columna
df['belongs_to_collection'] = df['belongs_to_collection'].apply(str_to_dict)


In [75]:
#una vez como diccionario podemos aplicar pd.json_normalize para desanidar
df = pd.concat([pd.json_normalize(df['belongs_to_collection']), df.drop(['belongs_to_collection'], axis=1)], axis=1)

In [76]:
df

id                            name  \
0       10194.0            Toy Story Collection   
1           NaN                             NaN   
2      119050.0       Grumpy Old Men Collection   
3           NaN                             NaN   
4       96871.0  Father of the Bride Collection   
...         ...                             ...   
45460       NaN                             NaN   
45462       NaN                             NaN   
45463       NaN                             NaN   
45464       NaN                             NaN   
45465       NaN                             NaN   

                            poster_path                     backdrop_path  \
0      /7G9915LfUQ2lVfwMEEhDsn3kT4B.jpg  /9FBwqcd9IRruEDUrTdcaafOMKUq.jpg   
1                                   NaN                               NaN   
2      /nLvUdqgPgm3F85NMCii9gVFUcet.jpg  /hypTnLot2z8wpFS7qwsQHW1uV8u.jpg   
3                                   NaN                               NaN   
4      /nts4iOmNnq7GNicycMJ9pSAn204.jpg  /7qwE57OVZmMJChBpLEbJEmzUydk.jpg   
...                                 ...                               ...   
45460                               NaN                               NaN   
45462                               NaN                               NaN   
45463                               NaN                               NaN   
45464                               NaN                               NaN   
45465                               NaN                               NaN   

           budget                                             genres      id  \
0      30000000.0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...     862   
1      65000000.0  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...    8844   
2             0.0  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...   15602   
3      16000000.0  [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...   31357   
4             0.0                     [{'id': 35, 'name': 'Comedy'}]   11862   
...           ...                                                ...     ...   
45460         0.0  [{'id': 18, 'name': 'Drama'}, {'id': 28, 'name...   30840   
45462         0.0                      [{'id': 18, 'name': 'Drama'}]  111109   
45463         0.0  [{'id': 28, 'name': 'Action'}, {'id': 18, 'nam...   67758   
45464         0.0                                                 []  227506   
45465         0.0                                                 []  461257   

      original_language popularity  \
0                    en  21.946943   
1                    en  17.015539   
2                    en    11.7129   
3                    en   3.859495   
4                    en   8.387519   
...                 ...        ...   
45460                en   5.683753   
45462                tl   0.178241   
45463                en   0.903007   
45464                en   0.003503   
45465                en   0.163015   

                                    production_companies  \
0         [{'name': 'Pixar Animation Studios', 'id': 3}]   
1      [{'name': 'TriStar Pictures', 'id': 559}, {'na...   
2      [{'name': 'Warner Bros.', 'id': 6194}, {'name'...   
3      [{'name': 'Twentieth Century Fox Film Corporat...   
4      [{'name': 'Sandollar Productions', 'id': 5842}...   
...                                                  ...   
45460  [{'name': 'Westdeutscher Rundfunk (WDR)', 'id'...   
45462             [{'name': 'Sine Olivia', 'id': 19653}]   
45463  [{'name': 'American World Pictures', 'id': 6165}]   
45464               [{'name': 'Yermoliev', 'id': 88753}]   
45465                                                 []   

                                    production_countries release_date  \
0      [{'iso_3166_1': 'US', 'name': 'United States o...   1995-10-30   
1      [{'iso_3166_1': 'US', 'name': 'United States o...   1995-12-15   
2      [{'iso_3166_1': 'US', 'name': 'United States o...   1995-12-22   
3      [{'iso_3166_1': 'US', 'name': 'United